oauth2client/contrib/gce.py

```diff
@@ -85,32 +138,128 @@ class AppAssertionCredentials(AssertionCredentials):
+        # will not pass the scopes kwarg to parent class
+        self._check_scopes_and_notify(scope)
 
         # Assertion type is no longer used, but still in the
         # parent class signature.
-        super(AppAssertionCredentials, self).__init__(None)
-        self._service_account_email = None
+        super(AppAssertionCredentials, self).__init__(
+            None,
+            scopes=scope,
+            **unused_kwargs
+        )
 
-    @classmethod
-    def from_json(cls, json_data):
-        data = json.loads(_from_bytes(json_data))
-        return AppAssertionCredentials(data['scope'])
+    @property
+    def service_account_info(self):
+        """Info about this service account.
+
+        This property is a deserialized JSON service account object of the form
+        {'aliases': [...], 'scopes': [...], 'email': 'a@example.com'}
+        where 'scopes' and 'email' will always be present.
+        """
+        return self._get_service_account_info()
+
+    @property
+    def scopes(self):
+        return self.service_account_info['scopes']
+
+    @scopes.setter
+    def scopes(self, value):
+        self._check_scopes_and_notify(value)
+
+    @property
+    def service_account_email(self):
+        return self.service_account_info['email']
+
+    @property
+    def project_id(self):
+        if not self._project_id:
+            self._project_id = _get_metadata(
+                path=['project', 'project-id'],
+                recursive=False
+            )
+        return self._project_id
+
+    @property
+    def serialization_data(self):
+        return {'service_account_info': self.service_account_info}
+
+    def _check_scopes_and_notify(self, scopes):
+        if scopes:
+            if self.has_scopes(scopes):
+                warnings.warn(_SCOPES_WARNING)
+            else:
+                raise AttributeError(_SCOPES_ERROR)
+
+    def _get_service_account_info(self, http_request=None):
+        """Retrieves the full info for a service account and caches it
+
+        Args:
+            http_request: an httplib2.Http().request object or equivalent
+                with which to make the call to the metadata server
+
+        Returns:
+            A deserialized JSON service account object of the form:
+                {
+                   'aliases': [...],
+                   'scopes': [...],
+                   'email': 'a@example.com'
+                }
+        """
+        if self._partial:
+            self._service_account_info = _get_metadata(
+                path=[
+                    'instance',
+                    'service-accounts',
+                    self._service_account_info['email'],
+                ],
+                http_request=http_request
+            )
+            self._partial = False
+        return self._service_account_info
+
+    def _retrieve_scopes(self, http_request):
+        return self._get_service_account_info(
+            http_request=http_request
+        )['scopes']
```

Comment:
I don't understand the line-wrap here.

Can the whole expression fit on one line?